# Task: Clean the dataset replace missing values, remove outliers etc.

In [79]:
# Importing the necessary libraries
import pandas as pd
import numpy as np

In [80]:
#Extracting Data from dataset
data=pd.read_csv("D:/CodeOnBytes/dataset - netflix1.csv")
data

show_id     type                             title         director  \
0         s1    Movie              Dick Johnson Is Dead  Kirsten Johnson   
1         s3  TV Show                         Ganglands  Julien Leclercq   
2         s6  TV Show                     Midnight Mass    Mike Flanagan   
3        s14    Movie  Confessions of an Invisible Girl    Bruno Garotti   
4         s8    Movie                           Sankofa     Haile Gerima   
...      ...      ...                               ...              ...   
8785   s8797  TV Show                        Yunus Emre        Not Given   
8786   s8798  TV Show                         Zak Storm        Not Given   
8787   s8801  TV Show                Zindagi Gulzar Hai        Not Given   
8788   s8784  TV Show                              Yoko        Not Given   
8789   s8786  TV Show                               YOM        Not Given   

            country  date_added  release_year rating   duration  \
0     United States   9/25/2021          2020  PG-13     90 min   
1            France   9/24/2021          2021  TV-MA   1 Season   
2     United States   9/24/2021          2021  TV-MA   1 Season   
3            Brazil   9/22/2021          2021  TV-PG     91 min   
4     United States   9/24/2021          1993  TV-MA    125 min   
...             ...         ...           ...    ...        ...   
8785         Turkey   1/17/2017          2016  TV-PG  2 Seasons   
8786  United States   9/13/2018          2016  TV-Y7  3 Seasons   
8787       Pakistan  12/15/2016          2012  TV-PG   1 Season   
8788       Pakistan   6/23/2018          2016   TV-Y   1 Season   
8789       Pakistan    6/7/2018          2016  TV-Y7   1 Season   

                                              listed_in  
0                                         Documentaries  
1     Crime TV Shows, International TV Shows, TV Act...  
2                    TV Dramas, TV Horror, TV Mysteries  
3                    Children & Family Movies, Comedies  
4      Dramas, Independent Movies, International Movies  
...                                                 ...  
8785                  International TV Shows, TV Dramas  
8786                                           Kids' TV  
8787  International TV Shows, Romantic TV Shows, TV ...  
8788                                           Kids' TV  
8789                                           Kids' TV  

[8790 rows x 10 columns]

In [82]:
#checking missing values
data.isna().sum()

show_id         0
type            0
title           0
director        0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
dtype: int64

Checking for empty values, Since missing values are of not np.nan we are checking for all other possibilities.

In [83]:
#show_id
count=0
for i in data.show_id:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [84]:
#type
count=0
for i in data.type:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [85]:
#title
count=0
for i in data.title:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [86]:
#director
count=0
for i in data.director:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

2588

In [87]:
#country
count=0
for i in data.country:
    if i=="Not Given":
       count+=1
count

287

In [88]:
#date added
count=0
for i in data.date_added:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [89]:
#release year
count=0
for i in data.release_year:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [90]:
#rating
count=0
for i in data.rating:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [91]:
#duration
count=0
for i in data.duration:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

In [92]:
#listed in
count=0
for i in data.listed_in:
    if i=="" or i==np.nan or i==" " or i=="Not Given":
       count+=1
count

0

# Handling Missing values

Since the feature data['director'] has a lot of missing values (close to 30%), it's best to remove that feature from the dataset.

In [93]:
data=data.drop(columns='director')
data

show_id     type                             title        country  \
0         s1    Movie              Dick Johnson Is Dead  United States   
1         s3  TV Show                         Ganglands         France   
2         s6  TV Show                     Midnight Mass  United States   
3        s14    Movie  Confessions of an Invisible Girl         Brazil   
4         s8    Movie                           Sankofa  United States   
...      ...      ...                               ...            ...   
8785   s8797  TV Show                        Yunus Emre         Turkey   
8786   s8798  TV Show                         Zak Storm  United States   
8787   s8801  TV Show                Zindagi Gulzar Hai       Pakistan   
8788   s8784  TV Show                              Yoko       Pakistan   
8789   s8786  TV Show                               YOM       Pakistan   

      date_added  release_year rating   duration  \
0      9/25/2021          2020  PG-13     90 min   
1      9/24/2021          2021  TV-MA   1 Season   
2      9/24/2021          2021  TV-MA   1 Season   
3      9/22/2021          2021  TV-PG     91 min   
4      9/24/2021          1993  TV-MA    125 min   
...          ...           ...    ...        ...   
8785   1/17/2017          2016  TV-PG  2 Seasons   
8786   9/13/2018          2016  TV-Y7  3 Seasons   
8787  12/15/2016          2012  TV-PG   1 Season   
8788   6/23/2018          2016   TV-Y   1 Season   
8789    6/7/2018          2016  TV-Y7   1 Season   

                                              listed_in  
0                                         Documentaries  
1     Crime TV Shows, International TV Shows, TV Act...  
2                    TV Dramas, TV Horror, TV Mysteries  
3                    Children & Family Movies, Comedies  
4      Dramas, Independent Movies, International Movies  
...                                                 ...  
8785                  International TV Shows, TV Dramas  
8786                                           Kids' TV  
8787  International TV Shows, Romantic TV Shows, TV ...  
8788                                           Kids' TV  
8789                                           Kids' TV  

[8790 rows x 9 columns]

Since, the 200 data points of df['country '] is missing, those missing data points alone will be removed.

In [94]:
data=data.drop(data[data['country']=='Not Given'].index)
data

show_id     type                             title        country  \
0         s1    Movie              Dick Johnson Is Dead  United States   
1         s3  TV Show                         Ganglands         France   
2         s6  TV Show                     Midnight Mass  United States   
3        s14    Movie  Confessions of an Invisible Girl         Brazil   
4         s8    Movie                           Sankofa  United States   
...      ...      ...                               ...            ...   
8785   s8797  TV Show                        Yunus Emre         Turkey   
8786   s8798  TV Show                         Zak Storm  United States   
8787   s8801  TV Show                Zindagi Gulzar Hai       Pakistan   
8788   s8784  TV Show                              Yoko       Pakistan   
8789   s8786  TV Show                               YOM       Pakistan   

      date_added  release_year rating   duration  \
0      9/25/2021          2020  PG-13     90 min   
1      9/24/2021          2021  TV-MA   1 Season   
2      9/24/2021          2021  TV-MA   1 Season   
3      9/22/2021          2021  TV-PG     91 min   
4      9/24/2021          1993  TV-MA    125 min   
...          ...           ...    ...        ...   
8785   1/17/2017          2016  TV-PG  2 Seasons   
8786   9/13/2018          2016  TV-Y7  3 Seasons   
8787  12/15/2016          2012  TV-PG   1 Season   
8788   6/23/2018          2016   TV-Y   1 Season   
8789    6/7/2018          2016  TV-Y7   1 Season   

                                              listed_in  
0                                         Documentaries  
1     Crime TV Shows, International TV Shows, TV Act...  
2                    TV Dramas, TV Horror, TV Mysteries  
3                    Children & Family Movies, Comedies  
4      Dramas, Independent Movies, International Movies  
...                                                 ...  
8785                  International TV Shows, TV Dramas  
8786                                           Kids' TV  
8787  International TV Shows, Romantic TV Shows, TV ...  
8788                                           Kids' TV  
8789                                           Kids' TV  

[8503 rows x 9 columns]

# Removing Outliers

In [95]:
data['country'].value_counts()
countries=dict()
s=0
for i in data.country:
    if i not in countries:
        countries[i]=1
    else:
        countries[i]+=1
    
countries

{'United States': 3240,
 'France': 213,
 'Brazil': 88,
 'United Kingdom': 638,
 'India': 1057,
 'Germany': 104,
 'Pakistan': 421,
 'China': 100,
 'South Africa': 42,
 'Japan': 259,
 'Nigeria': 105,
 'Spain': 182,
 'Philippines': 83,
 'Australia': 114,
 'Argentina': 76,
 'Canada': 271,
 'Hong Kong': 79,
 'Italy': 66,
 'New Zealand': 17,
 'Egypt': 123,
 'Colombia': 42,
 'Mexico': 138,
 'Belgium': 18,
 'Switzerland': 8,
 'Taiwan': 86,
 'Bulgaria': 5,
 'Poland': 30,
 'South Korea': 214,
 'Saudi Arabia': 12,
 'Thailand': 66,
 'Indonesia': 86,
 'Kuwait': 7,
 'Malaysia': 23,
 'Vietnam': 7,
 'Lebanon': 24,
 'Romania': 10,
 'Syria': 2,
 'United Arab Emirates': 21,
 'Sweden': 19,
 'Mauritius': 2,
 'Austria': 9,
 'Turkey': 112,
 'Czech Republic': 6,
 'Cameroon': 1,
 'Netherlands': 28,
 'Ireland': 28,
 'Russia': 19,
 'Kenya': 4,
 'Chile': 22,
 'Uruguay': 9,
 'Bangladesh': 3,
 'Portugal': 4,
 'Hungary': 5,
 'Norway': 21,
 'Singapore': 28,
 'Iceland': 6,
 'Serbia': 3,
 'Namibia': 1,
 'Peru': 6,
 'Mo

In [96]:
#Finding 10% outlier data to be removed
minimum=[]
Sum=0
for i in countries:
    if countries[i]>=1 and countries[i]<=80:
        minimum.append(i)
        Sum+=countries[i]
print(Sum,len(minimum))

869 65


In [97]:
for i in minimum:
    data=data.drop(data[data['country']==i].index)

In [98]:
data

show_id     type                             title        country  \
0         s1    Movie              Dick Johnson Is Dead  United States   
1         s3  TV Show                         Ganglands         France   
2         s6  TV Show                     Midnight Mass  United States   
3        s14    Movie  Confessions of an Invisible Girl         Brazil   
4         s8    Movie                           Sankofa  United States   
...      ...      ...                               ...            ...   
8785   s8797  TV Show                        Yunus Emre         Turkey   
8786   s8798  TV Show                         Zak Storm  United States   
8787   s8801  TV Show                Zindagi Gulzar Hai       Pakistan   
8788   s8784  TV Show                              Yoko       Pakistan   
8789   s8786  TV Show                               YOM       Pakistan   

      date_added  release_year rating   duration  \
0      9/25/2021          2020  PG-13     90 min   
1      9/24/2021          2021  TV-MA   1 Season   
2      9/24/2021          2021  TV-MA   1 Season   
3      9/22/2021          2021  TV-PG     91 min   
4      9/24/2021          1993  TV-MA    125 min   
...          ...           ...    ...        ...   
8785   1/17/2017          2016  TV-PG  2 Seasons   
8786   9/13/2018          2016  TV-Y7  3 Seasons   
8787  12/15/2016          2012  TV-PG   1 Season   
8788   6/23/2018          2016   TV-Y   1 Season   
8789    6/7/2018          2016  TV-Y7   1 Season   

                                              listed_in  
0                                         Documentaries  
1     Crime TV Shows, International TV Shows, TV Act...  
2                    TV Dramas, TV Horror, TV Mysteries  
3                    Children & Family Movies, Comedies  
4      Dramas, Independent Movies, International Movies  
...                                                 ...  
8785                  International TV Shows, TV Dramas  
8786                                           Kids' TV  
8787  International TV Shows, Romantic TV Shows, TV ...  
8788                                           Kids' TV  
8789                                           Kids' TV  

[7634 rows x 9 columns]

# Checking for corelation between features

The features which are possible to be labelled will be lablled using label encoder.

In [99]:
from sklearn.preprocessing import LabelEncoder
encode=LabelEncoder()

# labelling type feature
encode.fit(data.type)
new_type=encode.transform(data.type)
data.type= new_type

# labelling country feature
encode.fit(data.country)
new_country=encode.transform(data.country)
data.country= new_country

#labelling date feature
encode.fit(data.date_added)
new_date=encode.transform(data.date_added)
data.date_added= new_date

#labelling rate feature
encode.fit(data.rating)
new_rate=encode.transform(data.rating)
data.rating= new_rate

#labelling rate feature
encode.fit(data.duration)
new_dur=encode.transform(data.duration)
data.duration= new_dur

#labelling show_id feature
encode.fit(data.show_id)
new_id=encode.transform(data.show_id)
data.show_id=new_id

In [100]:
data

show_id  type                             title  country  date_added  \
0           0     0              Dick Johnson Is Dead       19        1610   
1        1899     1                         Ganglands        5        1605   
2        4814     1                     Midnight Mass       19        1605   
3         389     0  Confessions of an Invisible Girl        1        1596   
4        6736     0                           Sankofa       19        1605   
...       ...   ...                               ...      ...         ...   
8785     7509     1                        Yunus Emre       17          41   
8786     7510     1                         Zak Storm       19        1542   
8787     7515     1                Zindagi Gulzar Hai       12         422   
8788     7499     1                              Yoko       12        1175   
8789     7500     1                               YOM       12        1233   

      release_year  rating  duration  \
0             2020       4       201   
1             2021       8         0   
2             2021       8         0   
3             2021       9       202   
4             1993       8        31   
...            ...     ...       ...   
8785          2016       9       107   
8786          2016      11       129   
8787          2012       9         0   
8788          2016      10         0   
8789          2016      11         0   

                                              listed_in  
0                                         Documentaries  
1     Crime TV Shows, International TV Shows, TV Act...  
2                    TV Dramas, TV Horror, TV Mysteries  
3                    Children & Family Movies, Comedies  
4      Dramas, Independent Movies, International Movies  
...                                                 ...  
8785                  International TV Shows, TV Dramas  
8786                                           Kids' TV  
8787  International TV Shows, Romantic TV Shows, TV ...  
8788                                           Kids' TV  
8789                                           Kids' TV  

[7634 rows x 9 columns]

In [101]:
data.corr()

C:\Users\pongr\AppData\Local\Temp\ipykernel_15368\2627137660.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data.corr()


show_id      type   country  date_added  release_year  \
show_id       1.000000 -0.135564  0.079688   -0.022485     -0.241057   
type         -0.135564  1.000000  0.022593    0.042063      0.189293   
country       0.079688  0.022593  1.000000   -0.029282     -0.048431   
date_added   -0.022485  0.042063 -0.029282    1.000000      0.079759   
release_year -0.241057  0.189293 -0.048431    0.079759      1.000000   
rating       -0.132468  0.315454 -0.125118    0.035064      0.259053   
duration      0.097619 -0.376792  0.129911   -0.009236      0.008668   

                rating  duration  
show_id      -0.132468  0.097619  
type          0.315454 -0.376792  
country      -0.125118  0.129911  
date_added    0.035064 -0.009236  
release_year  0.259053  0.008668  
rating        1.000000 -0.029953  
duration     -0.029953  1.000000

To Sum it up:-
1) We try to find the missing values in the dataset. Since the dataset's missing values aren't of type 'np.nan', we use other possible methods to find them based on reviewing the dataset.
2) Missing values are handled by removing the feature 'director' because close to 30% of the feature is empty. Also, the feature country has about 287 empty data points.Since, 287 is less than 5% of missing values only those datapoints were removed.
3) The outliers were removed by removing the countries with low no of series/movies, accounting close to 10% of the data.
4) Finally, corelation between features were checked by labelling them. Since the correlation is low according to corr() we didn't take any step here.

Hence Data has been preprocessed.